core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala

```diff
@@ -0,0 +1,131 @@
+  ): CompletableFuture[JoinGroupResponseData] = {
+    val future = new CompletableFuture[JoinGroupResponseData]()
+
+    def callback(joinResult: JoinGroupResult): Unit = {
+      future.complete(new JoinGroupResponseData()
+        .setErrorCode(joinResult.error.code)
+        .setGenerationId(joinResult.generationId)
+        .setProtocolType(joinResult.protocolType.orNull)
+        .setProtocolName(joinResult.protocolName.orNull)
+        .setLeader(joinResult.leaderId)
+        .setSkipAssignment(joinResult.skipAssignment)
+        .setMemberId(joinResult.memberId)
+        .setMembers(joinResult.members.asJava)
+      )
+    }
+
+    val groupInstanceId = Option(request.groupInstanceId)
+
+    // Only return MEMBER_ID_REQUIRED error if joinGroupRequest version is >= 4
+    // and groupInstanceId is configured to unknown.
+    val requireKnownMemberId = context.apiVersion >= 4 && groupInstanceId.isEmpty
+
+    val protocols = request.protocols.valuesList.asScala.map { protocol =>
+      (protocol.name, protocol.metadata)
+    }.toList
+
+    val supportSkippingAssignment = context.apiVersion >= 9
+
+    coordinator.handleJoinGroup(
+      request.groupId,
+      request.memberId,
+      groupInstanceId,
+      requireKnownMemberId,
+      supportSkippingAssignment,
+      context.clientId,
+      context.clientAddress.toString,
+      request.rebalanceTimeoutMs,
+      request.sessionTimeoutMs,
+      request.protocolType,
+      protocols,
+      callback,
+      Option(request.reason),
+      RequestLocal(bufferSupplier)
+    )
+
+    future
+  }
+
+  override def deleteOffsets(
+    context: RequestContext,
+    request: OffsetDeleteRequestData,
+    bufferSupplier: BufferSupplier
+  ): CompletableFuture[OffsetDeleteResponseData] = {
+    val future = new CompletableFuture[OffsetDeleteResponseData]()
+
+    val partitions = mutable.ArrayBuffer[TopicPartition]()
+    request.topics.forEach { topic =>
+      topic.partitions.forEach { partition =>
+        partitions += new TopicPartition(topic.name, partition.partitionIndex)
+      }
+    }
+
+    val (groupError, topicPartitionResults) = coordinator.handleDeleteOffsets(
+      request.groupId,
+      partitions,
+      RequestLocal(bufferSupplier)
+    )
+
+    if (groupError != Errors.NONE) {
+      future.completeExceptionally(groupError.exception)
+    } else {
+      val response = new OffsetDeleteResponseData()
+      topicPartitionResults.forKeyValue { (topicPartition, error) =>
+        var topic = response.topics.find(topicPartition.topic)
+        if (topic == null) {
+          topic = new OffsetDeleteResponseData.OffsetDeleteResponseTopic().setName(topicPartition.topic)
+          response.topics.add(topic)
+        }
+        topic.partitions.add(new OffsetDeleteResponseData.OffsetDeleteResponsePartition()
+          .setPartitionIndex(topicPartition.partition)
+          .setErrorCode(error.code))
+      }
+
+      future.complete(response)
+    }
+
+    future
```

Comment:
Looks like we're moving to a different async model, is there a design doc to read about this?

Reply:
KIP-848 is the source of truth. The KIP explains that we want to move to a event loop based group coordinator. This is why all the methods are using futures.

Reply:
From the KIP-848: ```This KIP proposes to evolve the group coordinator to rely on an event loop. The rational of using an event loop is that 1) it simplifies the concurrency and 2) enables simulation testing. The group coordinator will have a replicated state machine per __consumer_offsets partitions, where each replicated state machine is modelled as an event loop. Those replicated state machines will be executed in group.coordinator.threads threads.``` 

Not sure I can easily see from that intent that we need to use futures.  E.g. request threads use event loop without futures, so I'm curious how the event loop with futures is going to look like (this is not to say that it's hard to do, but looks like a new concept in Kafka, so it would be good to see a more detailed design,).

Reply:
The concept is not new in Kafka. This is how we implemented the quorum controller. I follow the same patterns that we have introduced there. I agree that future are not strictly necessary. We could keep using callbacks as today. I find Future nicer to work with, personally.

I don't have a detailed design doc for this part as we usually treat this as implementation details. However, I will have a first implementation of the core building block of the new group coordinator soon. This will help to understand the new architecture. Does it work for you?

Reply:
Sounds good.